File: js_solutions/binary-search.js
const binarySearch = (nums, target) => {
  let start = 0;
  let end = nums.length - 1;

  while (start <= end) {
    const midIndex = Math.floor(start + (end - start) / 2);
    const midElement = nums[midIndex];
    if (midElement === target) return midIndex;

    if (midElement < target) {
      start = midIndex + 1;
    } else {
      end = midIndex - 1;
    }
  }
  return -1;
}

File: 01_session/easy/binary_search.rb
# frozen_string_literal: true

# using builtin
def search(nums, target)
  nums.bsearch_index { |num| target <=> num } || -1
end

# hand-rolled bsearch
def hand_made_search(nums, target)
  lo = 0
  hi = nums.length - 1

  while lo <= hi
    mid_idx = (lo + ((hi - lo) / 2)).floor
    mid_element = nums[mid_idx]
    return mid_idx if mid_element == target

    if mid_element < target
      lo = mid_idx + 1
    else
      hi = mid_idx - 1
    end
  end
  -1
end
